alibi_detect/od/pytorch/ensemble.py

```diff
@@ -0,0 +1,341 @@
+        p_vals = (
+                1 + (scores[:, None, :] < self.val_scores[None, :, :]).sum(1)
+            )/(len(self.val_scores)+1)
+        return 1 - p_vals
+
+
+class ShiftAndScaleNormalizer(BaseFittedTransformTorch):
+    def __init__(self):
+        """Maps scores to their normalised values.
+
+        Needs to be fit (see :py:obj:`~alibi_detect.od.pytorch.ensemble.BaseFittedTransformTorch`).
+        Subtracts the dataset mean and scales by the standard deviation.
+        """
+        super().__init__()
+        self.val_means = None
+        self.val_scales = None
+
+    def _fit(self, val_scores: torch.Tensor) -> ShiftAndScaleNormalizer:
+        """Computes the mean and standard deviation of the scores and stores them.
+
+        Parameters
+        ----------
+        val_scores
+            `Torch.Tensor` of scores from ensemble of detectors.
+        """
+        self.val_means = val_scores.mean(0)[None, :]
+        self.val_scales = val_scores.std(0)[None, :]
+        return self
+
+    def _transform(self, scores: torch.Tensor) -> torch.Tensor:
+        """Transform scores to normalised values. Subtracts the mean and scales by the standard deviation.
+
+        Parameters
+        ----------
+        scores
+            `Torch.Tensor` of scores from ensemble of detectors.
+
+        Returns
+        -------
+        `Torch.Tensor` of normalised scores.
+        """
+        return (scores - self.val_means)/self.val_scales
+
+
+class TopKAggregator(BaseTransformTorch):
+    def __init__(self, k: Optional[int] = None):
+        """Takes the mean of the top `k` scores.
+
+        Parameters
+        ----------
+        k
+            number of scores to take the mean of. If `k` is left ``None`` then will be set to
+            half the number of scores passed in the forward call.
+        """
+        super().__init__()
+        self.k = k
+
+    def _transform(self, scores: torch.Tensor) -> torch.Tensor:
+        """Takes the mean of the top `k` scores.
+
+        Parameters
+        ----------
+        scores
+            `Torch.Tensor` of scores from ensemble of detectors.
+
+        Returns
+        -------
+        `Torch.Tensor` of mean of top `k` scores.
+        """
+        if self.k is None:
+            self.k = int(np.ceil(scores.shape[1]/2))
+        sorted_scores, _ = torch.sort(scores, 1)
+        return sorted_scores[:, -self.k:].mean(-1)
+
+
+class AverageAggregator(BaseTransformTorch):
+    def __init__(self, weights: Optional[torch.Tensor] = None):
+        """Averages the scores of the detectors in an ensemble.
+
+        Parameters
+        ----------
+        weights
+            Optional parameter to weight the scores. If `weights` is left ``None`` then will be set to
+            a vector of ones.
+
+        Raises
+        ------
+        ValueError
+            If `weights` does not sum to ``1``.
+        """
+        super().__init__()
+        if weights is not None and weights.sum() != 1:
```

Comment:
Is this likely to fail due to floating point operations?

Reply:
I've added `np.isclose` check instead of using `weights.sum() != 1`